vault/rekey.go

```diff
@@ -757,168 +736,98 @@ func (c *Core) BarrierRekeyVerify(ctx context.Context, key []byte, nonce string)
 	c.rekeyLock.Lock()
 	defer c.rekeyLock.Unlock()
 
+	config := c.barrierRekeyConfig
+	if recovery {
+		config = c.recoveryRekeyConfig
+	}
+
 	// Ensure a rekey is in progress
-	if c.barrierRekeyConfig == nil {
-		return nil, logical.CodedError(http.StatusBadRequest, "no barrier rekey in progress")
+	if config == nil {
+		return nil, logical.CodedError(http.StatusBadRequest, "no rekey in progress")
 	}
 
-	if nonce != c.barrierRekeyConfig.VerificationNonce {
-		return nil, logical.CodedError(http.StatusBadRequest, fmt.Sprintf("incorrect nonce supplied; nonce for this verify operation is %q", c.barrierRekeyConfig.VerificationNonce))
+	if nonce != config.VerificationNonce {
+		return nil, logical.CodedError(http.StatusBadRequest, fmt.Sprintf("incorrect nonce supplied; nonce for this verify operation is %q", config.VerificationNonce))
 	}
 
 	// Check if we already have this piece
-	for _, existing := range c.barrierRekeyVerifyProgress {
+	for _, existing := range config.VerificationProgress {
 		if subtle.ConstantTimeCompare(existing, key) == 1 {
 			return nil, logical.CodedError(http.StatusBadRequest, "given key has already been provided during this verify operation")
 		}
 	}
 
 	// Store this key
-	c.barrierRekeyVerifyProgress = append(c.barrierRekeyVerifyProgress, key)
+	config.VerificationProgress = append(config.VerificationProgress, key)
 
 	// Check if we don't have enough keys to unlock
-	if len(c.barrierRekeyVerifyProgress) < c.barrierRekeyConfig.SecretThreshold {
+	if len(config.VerificationProgress) < config.SecretThreshold {
 		if c.logger.IsDebug() {
-			c.logger.Debug("cannot verify yet, not enough keys", "keys", len(c.barrierRekeyVerifyProgress), "threshold", c.barrierRekeyConfig.SecretThreshold)
+			c.logger.Debug("cannot verify yet, not enough keys", "keys", len(config.VerificationProgress), "threshold", config.SecretThreshold)
 		}
 		return nil, nil
 	}
 
 	// Schedule the progress for forgetting and rotate the nonce if possible
 	defer func() {
-		c.barrierRekeyVerifyProgress = nil
-		if c.barrierRekeyConfig != nil {
-			nonce, err := uuid.GenerateUUID()
-			if err == nil {
-				c.barrierRekeyConfig.VerificationNonce = nonce
-				if ret != nil {
-					ret.Nonce = nonce
+		config.VerificationProgress = nil
+		nonce, err := uuid.GenerateUUID()
+		switch recovery {
+		case false:
+			if c.barrierRekeyConfig != nil {
+				if err == nil {
+					c.barrierRekeyConfig.VerificationNonce = nonce
+					if ret != nil {
+						ret.Nonce = nonce
+					}
+				}
+			}
+		default:
```

Comment:
Might be missing it, but i don't see a difference between these two cases.

Reply:
I think one should be on recoveryConfig.

Reply:
I also think this whole switch could be removed and you can set the value on config?

Reply:
Agreed -- even if we set the core object to nil, updating this here won't be a problem.